### ISSUE TITLE
Do not warn unreachable code for variable declarations

### ISSUE THREAD

**Issue by mttcr** [google/closure-compiler#2240]
From https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Statements/var :

> var hoisting 
> Because variable declarations (and declarations in general) are processed before any code is executed, declaring a variable anywhere in the code is equivalent to declaring it at the top. This also means that a variable can appear to be used before it's declared. This behavior is called "hoisting", as it appears that the variable declaration is moved to the top of the function or global code.

If there is a return statement in a function, variable declarations written after should not warn for unreachable code.
I did not intentionnally write a statement like that, it's from typescript transpilation:

```ts
export class A
{
    public a: A;
    private v( ...params: string[] )
    {

    }
    public test()
    {
        return this.a.v( ...[] );
    }
}
```
Is transpiled in:
```js
var A = (function () {
    function A() {
    }
    A.prototype.v = function () {
        var params = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            params[_i] = arguments[_i];
        }
    };
    A.prototype.test = function () {
        return (_a = this.a).v.apply(_a, []);
        var _a;
    };
    return A;
}());
exports.A = A;
```

**Comment by MatrixFrog** [comment]
@mprobst do you have tools that shift things around so that the `var` statement comes before the `return`, or something similar that solves this?

**Comment by mprobst** [comment]
@evmar any intuition on how hard this'd be to solve in tsickle (e.g. assuming we generate a duplicate declarations at the top of the file)?

**Comment by evmar** [comment]
I experimented a bit.

Even this code:

```
    let x = this.a.v( ...[] );
    return x;
```

is transpiled as

```
       var x = (_a = this.a).v.apply(_a, []);
        return x;
        var _a;
```

We currently do a pass over the compilation output (to rewrite `require` statements as `goog.module`), so it's possible for us to transform this as well.  It'd definitely be inconvenient; we're not transforming AST so much as moving text around.  On the other hand I dunno if it makes sense for Closure compiler to accept this code because it's pretty suspicious looking.

**Comment by Dominator008** [comment]
Cross referencing https://github.com/Microsoft/TypeScript/issues/7017

I thought TypeScript fixed this in 2.0 but apparently not.

**Comment by ChadKillingsworth** [comment]
I would prefer we NOT accept such code. Encouraging users to take advantage of hoisting with variables and functions defined after an exit point is not something I want to see.
